File: backend/src/main/java/com/bsharat/triosuite/invoice/InvoiceCalculator.java
package com.bsharat.triosuite.invoice;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import org.springframework.stereotype.Component;

/**
 * The single source of truth for invoice arithmetic.
 *
 * <p>Pure and stateless: every input arrives as an argument and nothing is cached, so the class can
 * be unit-tested with {@code new InvoiceCalculator()} and no Spring context. The server calls it on
 * every create and every update and stores what it returns — totals sent by a client are read for
 * nothing but comparison and are otherwise ignored.
 *
 * <p><b>Rounding.</b> {@link RoundingMode#HALF_UP} applied <em>per line</em>, to the invoice
 * currency's minor units (2 for ILS/USD/EUR/GBP, 3 for JOD). Invoice totals are the sum of the
 * already-rounded line amounts, never a re-rounding of an unrounded sum, so
 * {@code Σ lines == invoice total} holds exactly. See ADR-0003.
 *
 * <p><b>Formulas.</b> With {@code q} = quantity, {@code p} = unit price and {@code r} = tax rate:
 * <ul>
 *   <li>{@link TaxMode#EXCLUSIVE}: {@code net = round(q·p)}, {@code tax = round(net·r)},
 *       {@code gross = net + tax}</li>
 *   <li>{@link TaxMode#INCLUSIVE}: {@code gross = round(q·p)}, {@code net = round(gross / (1 + r))},
 *       {@code tax = gross − net}</li>
 * </ul>
 * In both modes {@code gross} is derived by addition or subtraction rather than by a second
 * rounding, which is what keeps {@code net + tax == gross} true for every line.
 */
@Component
public class InvoiceCalculator {

    /** Intermediate scale for the inclusive-mode division, well beyond any currency's precision. */
    private static final int DIVISION_SCALE = 12;

    /**
     * One line as submitted by a client, before any arithmetic.
     *
     * @param quantity  units sold; must be greater than zero
     * @param unitPrice price per unit, net or gross depending on the invoice's tax mode
     * @param taxRate   tax rate as a decimal fraction, e.g. {@code 0.1600} for 16%
     */
    public record LineInput(BigDecimal quantity, BigDecimal unitPrice, BigDecimal taxRate) {

        public LineInput {
            Objects.requireNonNull(quantity, "quantity");
            Objects.requireNonNull(unitPrice, "unitPrice");
            Objects.requireNonNull(taxRate, "taxRate");
        }
    }

    /**
     * A line after rounding. {@code netAmount + taxAmount == grossAmount} always holds.
     *
     * @param netAmount   amount excluding tax
     * @param taxAmount   tax on this line
     * @param grossAmount amount including tax
     */
    public record CalculatedLine(BigDecimal netAmount, BigDecimal taxAmount, BigDecimal grossAmount) {
    }

    /**
     * A whole invoice after rounding.
     *
     * @param lines          per-line results, positionally matching the input
     * @param subtotal       Σ net
     * @param taxTotal       Σ tax
     * @param grandTotal     Σ gross
     * @param grandTotalBase {@code grandTotal × exchangeRate}, rounded to the base currency
     */
    public record CalculatedInvoice(
            List<CalculatedLine> lines,
            BigDecimal subtotal,
            BigDecimal taxTotal,
            BigDecimal grandTotal,
            BigDecimal grandTotalBase) {

        public CalculatedInvoice {
            lines = List.copyOf(lines);
        }
    }

    /**
     * Computes one line.
     *
     * @param line        quantity, unit price and tax rate
     * @param taxMode     whether {@code unitPrice} is net or gross
     * @param minorUnits  decimal places of the invoice currency
     * @return the rounded net, tax and gross amounts
     */
    public CalculatedLine calculateLine(LineInput line, TaxMode taxMode, int minorUnits) {
        Objects.requireNonNull(line, "line");
        Objects.requireNonNull(taxMode, "taxMode");

        BigDecimal extended = line.quantity().multiply(line.unitPrice());
        BigDecimal net;
        BigDecimal tax;
        BigDecimal gross;

        if (taxMode == TaxMode.EXCLUSIVE) {
            net = extended.setScale(minorUnits, RoundingMode.HALF_UP);
            tax = net.multiply(line.taxRate()).setScale(minorUnits, RoundingMode.HALF_UP);
            gross = net.add(tax);
        } else {
            gross = extended.setScale(minorUnits, RoundingMode.HALF_UP);
            BigDecimal divisor = BigDecimal.ONE.add(line.taxRate());
            net = gross.divide(divisor, DIVISION_SCALE, RoundingMode.HALF_UP)
                    .setScale(minorUnits, RoundingMode.HALF_UP);
            tax = gross.subtract(net);
        }

        return new CalculatedLine(net, tax, gross);
    }

    /**
     * Computes every line and the invoice totals.
     *
     * @param lines                  the lines to price; may be empty, which yields zero totals
     * @param taxMode                whether unit prices are net or gross
     * @param currencyMinorUnits     decimal places of the invoice currency
     * @param exchangeRate           base-currency units per one invoice-currency unit
     * @param baseCurrencyMinorUnits decimal places of the base (reporting) currency
     * @return the rounded lines and totals
     */
    public CalculatedInvoice calculate(
            List<LineInput> lines,
            TaxMode taxMode,
            int currencyMinorUnits,
            BigDecimal exchangeRate,
            int baseCurrencyMinorUnits) {

        Objects.requireNonNull(lines, "lines");
        Objects.requireNonNull(exchangeRate, "exchangeRate");

        List<CalculatedLine> calculated = new ArrayList<>(lines.size());
        BigDecimal subtotal = zero(currencyMinorUnits);
        BigDecimal taxTotal = zero(currencyMinorUnits);
        BigDecimal grandTotal = zero(currencyMinorUnits);

        for (LineInput line : lines) {
            CalculatedLine result = calculateLine(line, taxMode, currencyMinorUnits);
            calculated.add(result);
            subtotal = subtotal.add(result.netAmount());
            taxTotal = taxTotal.add(result.taxAmount());
            grandTotal = grandTotal.add(result.grossAmount());
        }

        BigDecimal grandTotalBase = grandTotal.multiply(exchangeRate)
                .setScale(baseCurrencyMinorUnits, RoundingMode.HALF_UP);

        return new CalculatedInvoice(calculated, subtotal, taxTotal, grandTotal, grandTotalBase);
    }

    private static BigDecimal zero(int minorUnits) {
        return BigDecimal.ZERO.setScale(minorUnits, RoundingMode.UNNECESSARY);
    }
}

File: mobile/lib/core/money/invoice_calculator.dart
import 'package:decimal/decimal.dart';

import 'tax_mode.dart';

/// One line as the user has entered it, before any arithmetic.
class LineInput {
  const LineInput({
    required this.quantity,
    required this.unitPrice,
    required this.taxRate,
  });

  final Decimal quantity;
  final Decimal unitPrice;
  final Decimal taxRate;
}

/// A line after rounding. `netAmount + taxAmount == grossAmount` always holds.
class CalculatedLine {
  const CalculatedLine({
    required this.netAmount,
    required this.taxAmount,
    required this.grossAmount,
  });

  final Decimal netAmount;
  final Decimal taxAmount;
  final Decimal grossAmount;
}

/// A whole invoice after rounding.
class CalculatedInvoice {
  const CalculatedInvoice({
    required this.lines,
    required this.subtotal,
    required this.taxTotal,
    required this.grandTotal,
    required this.grandTotalBase,
  });

  final List<CalculatedLine> lines;
  final Decimal subtotal;
  final Decimal taxTotal;
  final Decimal grandTotal;

  /// [grandTotal] converted at the invoice's exchange rate, rounded to the base currency.
  final Decimal grandTotalBase;
}

/// The on-device mirror of the server's `InvoiceCalculator`.
///
/// This exists for one reason: so the totals footer on the Create screen can update as the user
/// types, without a round trip per keystroke. It is a **preview**. The server recomputes everything
/// on save and its answer is what gets stored and displayed afterwards — this class never decides
/// what an invoice is worth.
///
/// That makes it all the more important that the two agree, because a preview that disagrees with
/// the saved invoice by a cent is worse than no preview at all. The formulas and the rounding rule
/// below are therefore identical to the Java implementation, line for line:
///
/// * **Exclusive**: `net = round(q·p)`, `tax = round(net·r)`, `gross = net + tax`
/// * **Inclusive**: `gross = round(q·p)`, `net = round(gross / (1 + r))`, `tax = gross − net`
///
/// Rounding is half-up, applied **per line**, to the invoice currency's minor units. Invoice totals
/// are the sum of the already-rounded lines, never a re-rounding of an unrounded sum. `Decimal`
/// keeps every value exact — no monetary amount in this app ever passes through a `double`.
///
/// `test/core/invoice_calculator_test.dart` re-checks the figures from the Java suite against this
/// implementation, so a change to one that is not made to the other fails the build.
abstract final class InvoiceCalculator {
  const InvoiceCalculator._();

  /// Working precision for the inclusive-mode division, far beyond any currency's scale.
  static const int _divisionScale = 12;

  /// Computes one line.
  static CalculatedLine line(
    LineInput input, {
    required TaxMode taxMode,
    required int minorUnits,
  }) {
    final extended = input.quantity * input.unitPrice;

    if (taxMode == TaxMode.exclusive) {
      final net = _round(extended, minorUnits);
      final tax = _round(net * input.taxRate, minorUnits);
      return CalculatedLine(
        netAmount: net,
        taxAmount: tax,
        grossAmount: net + tax,
      );
    }

    final gross = _round(extended, minorUnits);
    final divisor = Decimal.one + input.taxRate;
    final net = _round(
      (gross / divisor).toDecimal(scaleOnInfinitePrecision: _divisionScale),
      minorUnits,
    );
    return CalculatedLine(
      netAmount: net,
      taxAmount: gross - net,
      grossAmount: gross,
    );
  }

  /// Computes every line and the invoice totals.
  ///
  /// [lines] may be empty, which yields zero totals rather than an error — an invoice with no
  /// lines is a perfectly valid draft.
  static CalculatedInvoice invoice({
    required List<LineInput> lines,
    required TaxMode taxMode,
    required int currencyMinorUnits,
    required Decimal exchangeRate,
    required int baseCurrencyMinorUnits,
  }) {
    final calculated = <CalculatedLine>[];
    var subtotal = Decimal.zero;
    var taxTotal = Decimal.zero;
    var grandTotal = Decimal.zero;

    for (final input in lines) {
      final result = line(input, taxMode: taxMode, minorUnits: currencyMinorUnits);
      calculated.add(result);
      subtotal += result.netAmount;
      taxTotal += result.taxAmount;
      grandTotal += result.grossAmount;
    }

    return CalculatedInvoice(
      lines: List.unmodifiable(calculated),
      subtotal: subtotal,
      taxTotal: taxTotal,
      grandTotal: grandTotal,
      grandTotalBase: _round(grandTotal * exchangeRate, baseCurrencyMinorUnits),
    );
  }

  /// Half-up to [scale] decimal places.
  ///
  /// `Decimal.round` rounds half away from zero, which is the same thing for the non-negative
  /// amounts this app deals in — quantities are validated as positive and prices as non-negative,
  /// on both sides of the wire.
  static Decimal _round(Decimal value, int scale) => value.round(scale: scale);
}
